akka-projection-grpc-tests/src/test/scala/akka/projection/grpc/internal/FilterStageSpec.scala

```diff
@@ -149,6 +154,45 @@ class FilterStageSpec extends ScalaTestWithActorTestKit("""
       outFuture.futureValue shouldBe envelopes.filterNot(_.persistenceId == PersistenceId(entityType, "b").id)
     }
 
+    "apply envelope filter after declarative filters excludes" in {
+      val envPublisherPromise = Promise[TestPublisher.Probe[EventEnvelope[Any]]]()
+      val envFilterProbe = createTestProbe[Offset]()
+      val envSource: Source[EventEnvelope[Any], _] =
+        TestSource()
+          .mapMaterializedValue(envPublisherPromise.success)
+      val envFlow: Flow[StreamIn, EventEnvelope[Any], NotUsed] =
+        BidiFlow
+          .fromGraph(
+            new FilterStage(
+              streamId,
+              entityType,
+              0 until persistence.numberOfSlices,
+              // exclude everything
+              Seq(FilterCriteria(FilterCriteria.Message.ExcludeMatchingEntityIds(ExcludeRegexEntityIds(Seq(".*"))))),
+              notUsedCurrentEventsByPersistenceIdQuery,
+              envelopeFilter = { envelope =>
+                envFilterProbe.ref ! envelope.offset
+                envelope.tags.contains("replicate-it")
+              }))
+          .join(Flow.fromSinkAndSource(Sink.ignore, envSource))
+      val streamIn: Source[StreamIn, TestPublisher.Probe[StreamIn]] = TestSource()
+
+      val (_, outFuture) = streamIn.via(envFlow).toMat(Sink.seq)(Keep.both).run()
+      val envPublisher = envPublisherPromise.future.futureValue
```

Comment:
I added `Setup` class to remove repeated code. That has been merged to dev-filter.